File: datasets/C++/Deepseek/Algorithm/hard5.cpp
#include <bits/stdc++.h>
using namespace std;
const int MOD = 1e9 + 7;

long long pow_mod(long long base, long long exp) {
    long long res = 1;
    while (exp > 0) {
        if (exp % 2 == 1) {
            res = (res * base) % MOD;
        }
        base = (base * base) % MOD;
        exp /= 2;
    }
    return res;
}

void solve() {
    int n, m;
    cin >> n >> m;
    int max_num = 1 << m;
    if (n > max_num - 1) {
        cout << 0 << endl;
        return;
    }
    
    vector<long long> fact(max_num);
    fact[0] = 1;
    for (int i = 1; i < max_num; ++i) {
        fact[i] = fact[i-1] * i % MOD;
    }
    
    long long ans = fact[max_num - 1] * pow_mod(fact[max_num - 1 - n], MOD - 2) % MOD;
    cout << ans << endl;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    solve();
    return 0;
}

File: datasets/Python/Deepseek/Algorithm/hard5.py
import sys
MOD = 10**9 + 7

def solve():
    n, m = map(int, sys.stdin.readline().split())
    max_num = 1 << m
    if n > max_num - 1:
        print(0)
        return
    
    # Precompute factorial modulo MOD up to max_num - 1
    fact = [1] * (max_num)
    for i in range(1, max_num):
        fact[i] = fact[i-1] * i % MOD
    
    # The answer is P(max_num - 1, n) since all configurations are winning when n=2
    ans = fact[max_num - 1] * pow(fact[max_num - 1 - n], MOD - 2, MOD) % MOD
    print(ans)

solve()